### DISCUSSION TITLE
Z-Offset Tool Not Working - 1.3.0

### DISCUSSION THREAD

**Discussion by klooney84** [Jyers/Marlin#505] [Q&A] [answered]
I am having an issue with the z-offset tool on v1.3.0 on the 4.2.7 board. I have flashed the 1.3.0 firmware 3 separate times but hasn't fixed the issue. No matter what I set the z-offset to, the z-offset does not change upon a new value and/or after a re-home or home. I've have saved the new offset to settings each time it was manipulated in addition to the auto save through the tool. I flashed to the older version (1.2.2) and the z-offset tool is now working.

**Comment by Jyers** [comment]
Have you tried going to control and selection reset to defaults and then store configuration? This process clears your eeprom which is very important after flashing. Not doing this tends to lead to all kinds of issues.

**Reply by klooney84** [reply]
I'll give it a shot and see if it helps. I reset to defaults but didn't store configuration when I was meeting with it about a week ago

**Comment by Schnabulation** [comment]
I can confirm this issue: Have been using the Z-offset tool successfully on earlier builds but with 1.3 the Z-offset does not stick!

Procedure I used successfully before:

1. Auto-Level with BL Touch
2. Enabled "Live Adjustment"
3. Set Z-offset to the correct value
4. Stored to EEPROM
5. Auto-Level again, then enabled "Live Adjustment" --> Z-offset set correctly.

Now with 1.3 it seems that the Z-offset value does not stick - neither when using "Live Adjustment" nor when doing a print (tested with a live print without filament inserted but with a paper underneath).

**Reply by Schnabulation** [reply]
EDIT: I always reset the EEPROM after loading a new firmware btw. Did not help at all.

**Reply by Schnabulation** [reply]
After further inspection it looks like "Live Adjustment" home Z not at 0.00 but at 0.10. Going back to V1.2.1 fixed it for me.

**Reply by Jyers** [reply]
Could not recreate this behavior. I tried everything I could to trip it up. I set the offset then rehomed, tried leveling before and after, tried saving and power cycling the printer, and no matter what I did the offset stuck. Nothing has changed about how the Z Offset stores or handles the offset value, so there is no difference between older versions and 1.3. I would encourage you to use a terminal to check your z offset to see if perhaps it isn't actually getting set for some reason.

**Reply by Schnabulation** [reply]
This is wierd... I seriously spend around 2 hours to try to narrow it down and the only thing that worked is going back to before you merged the latest changes from upstream bugfix-2.0.x branch. I was using the Z-offset tool heavily before so I would say I knew how it worked.

When I did it with 1.3.0 I could save the Z-probe offset and it would be stored to EEPROM (verified with `M503`) but when I went back into the Z-offset menu and enabled 'Live Adjustment' it would home and then move the nozzle down to only 0.10. When I moved the nozzle down to 0.00 with the 'Move' menu then the offset would be correct. So I suspect that the Z-offset gets saved correctly but 'Live Adjustment' has some kind of bug...

**Reply by Schnabulation** [reply]
Furthermore: I think you have "Z_SAFE_HOMING" enabled if I am not mistaken. Maybe there could be something related to this?

**Reply by Schnabulation** [reply]
I just compared `Marlin\src\lcd\extui\lib\creality_dwin\creality_dwin.cpp` from the working version with the latest version and I suspect something around the lines 3807 - 3819. Especially this part:

`else if (active_menu == Tune && selection == TUNE_ZOFFSET) {`
      `char buf[20];`
     `sprintf(buf, "M290 Z%f", (tempvalue/valueunit - zoffsetvalue));`
    `  gcode.process_subcommands_now_P(buf);`
  `  }`

It looks like in the latest version you do something with babystepping which is not present in the working version.
No idea if this is the problem to - I suck at programming.

**Reply by Jyers** [reply]
Nothing has changed about how live adjust gets initiated, and the gcode I'm sending tells the printer to go to z 0 after live adjust is enabled. The babystepping is just for during printing to fix a previous issue. Let me know if you figure anything else out, as at the moment I can not imagine why your printer would be going to z 0.1

**Reply by Schnabulation** [reply]
Will do - thanks for your help. I'll post here once I find something new...

**Comment by Schnabulation** [comment]
So, regarding this issue, I have made some progress:

It looks like when I enable "Live Adjustment" in the Z-offset tool it disables mesh leveling. I can confirm that with a terminal sending `M420`:

`Send: M420`
`Recv: echo:Bed Leveling OFF`
`Recv: echo:Fade Height 10.00`
`Recv: ok`

I have set `#define RESTORE_LEVELING_AFTER_G28` and can confirm that mesh leveling was active prior to enabling "Live Adjustment". Also, when I manually enable leveling again with `M420 S1` it seems that homing uses the correct offset.

**Reply by Schnabulation** [reply]
...and to add onto this: I think it may be related to the fact that you use the BL Touch to home Z. I don't, I use the existing endstop switch and home with this. So my home is actually different from the trigger point of the BL Touch, therefore I need mesh leveling to be enabled to live-adjust the Z-probe offset.

**Reply by Jyers** [reply]
That is quite interesting. The mesh is supposed to be disabled when setting the offset as having it enabled would in essence, offset your offset. I don't think this should be the issue, but if you find that it is actually the disabled mesh that is causing problems I will see what I can do to add a condition to avoid such behavior.

**Reply by Schnabulation** [reply]
I will double check by re-loading the 1.2.1 firmware where I don‘t have said problem and see, if mesh leveling stays enabled there. I‘ll report back...

**Reply by hunkin1969** [reply]
I think this depends also to a problem i found today. Im using BLTouch since beginning but never removed the mechanic Z-Stop. Yesterday evening i changed my Mobo and PSU cover due bigger fans and removed also the z-Stop. Now this morning i wanted to start printing. But Homing didnt worked. Z-Rod moved only up, but not down. Tried lot of things , even disassembled my covers again and measured vrefs, because i adjusted them yesterday also. They were much too high in factory setted. Last point i was thinking was the mechanic z-Stop. But this was the reason. As i put it back and wired the system works now fine again.

**Reply by Schnabulation** [reply]
> But Homing didnt worked. Z-Rod moved only up, but not down.

Did you enable `USE_PROBE_FOR_Z_HOMING` when you removed the Z-axis endstop?

**Reply by klooney84** [reply]
> I think this depends also to a problem i found today. Im using BLTouch since beginning but never removed the mechanic Z-Stop. Yesterday evening i changed my Mobo and PSU cover due bigger fans and removed also the z-Stop. Now this morning i wanted to start printing. But Homing didnt worked. Z-Rod moved only up, but not down. Tried lot of things , even disassembled my covers again and measured vrefs, because i adjusted them yesterday also. They were much too high in factory setted. Last point i was thinking was the mechanic z-Stop. But this was the reason. As i put it back and wired the system works now fine again.

This is exactly my issue. I have to open up the case and se if my z-stop is still plugged in or maybe I plugged the BLTouch into the Z-Switch port. It always worked with the Smith3D firmware, was caught off guard when I switched to jyers and the z-offset performed the same as you mentioned.

**Comment by Schnabulation** [comment, marked as answer]
I found the bug and created a new issue: https://github.com/Jyers/Marlin/issues/557
